### PR TITLE
Improve struct completions

### PR DESCRIPTION
**Pull request by scohen**
Two commits here. One improves struct completions, and the other makes them snippets to put the cursor inside the curly braces when you complete.

### CODE REVIEW

**Comment by scohen** [comment]
Fixes #122 and #123

File: apps/server/lib/lexical/server/code_intelligence/completion/translations/module_or_behaviour.ex [hunk 1]
```diff
@@ -15,45 +16,73 @@ defmodule Lexical.Server.CodeIntelligence.Completion.Translations.ModuleOrBehavi
...
+    end
+  end
 
-        true ->
-          {module.name, ""}
-      end
+  defp struct_completion(builder, %Env{} = env, module_name) do
+    last_word = Env.last_word(env)
+
+    Logger.info("last word is: #{last_word}")
```
Comment (scohen) [inline apps/server/lib/lexical/server/code_intelligence/completion/translations/module_or_behaviour.ex:43]: Remove

Comment (scohen) [inline apps/server/lib/lexical/server/code_intelligence/completion/translations/module_or_behaviour.ex:43]:
```suggestion
```

**Comment by scottming** [comment]
There is a bug when you want to complete the child module

```elixir
17:41:08.894 [info] last word is: %L
17:41:08.894 [info] add percent for exical.Project? true add curlies: true
17:41:08.895 [warn] Emitting completions: [#Protocol.Types.Completion.Item<[detail: "exical.Document (Struct)", insert_text: "%exical.Document{$1}", insert_text_format: :snippet, kind: :struct, label: "%exical.Document", sort_text: "exical.Document"]>, #Protocol.Types.Completion.Item<[detail: "exical.Project (Struct)", insert_text: "%exical.Project{$1}", insert_text_format: :snippet, kind: :struct, label: "%exical.Project", sort_text: "exical.Project"]>, #Protocol.Types.Completion.Item<[detail: "Linked lists hold zero, one, or more elements in the chosen order.", insert_text: "List", kind: :module, label: "List"]>, #Protocol.Types.Completion.Item<[detail: "A logger for Elixir applications.", insert_text: "Logger", kind: :module, label: "Logger"]>]
17:41:08.895 [info] Got [{"detail":"exical.Document (Struct)","insertText":"%exical.Document{$1}","insertTextFormat":2,"kind":22,"label":"%exical.Document","sortText":"exical.Document"},{"detail":"exical.Project (Struct)","insertText":"%exical.Project{$1}","insertTextFormat":2,"kind":22,"label":"%exical.Project","sortText":"exical.Project"},{"detail":"Linked lists hold zero, one, or more elements in the chosen order.","insertText":"List","kind":9,"label":"List"},{"detail":"A logger for Elixir applications.","insertText":"Logger","kind":9,"label":"Logger"}]
```

image

**Comment by scottming** [comment]
2. when you type `%Lexic` still no completions

image

3. type `%Lexical.D` still no module completions, only one struct completion(and this is not a snippet)

image

It should to complete both the struct and module.

4. when my hint is starts with `%`, the completions should not show me the functions

image






BTW, completions are complicated.

**Comment by scottming** [comment]
Another point to consider is whether you really need the cursor to be inside the braces when you are in the parameters of a function.

Because we may write a lot of `def func(%__MODULE__{} = module, _)` and `def handle_info(_, %State{}=state)` statements, compared to situations like `%Struct{|}`, seems like we use `%Struct{}|` more often when in parameters.

**Comment by scohen** [comment]
@scottming i can't fix the no completions case, if elixir sense doesn't give me completions, there's nothing I can do.

**Comment by scohen** [comment]
> Another point to consider is whether you really need the cursor to be inside the braces when you are in the parameters of a function

It's pretty common to pattern match on struct fields in function heads, and easy to tab out of the curly braces. I'd favor consistency here, since people will build up muscle memory about how to handle these things. Having a completion do one thing in one context and something else in another will be tough.

**Comment by scohen** [comment]
> type %Lexical.D still no module completions, only one struct completion(and this is not a snippet)

if there's a leading `%`, no module completions are offered, this is intentional, since you're in a struct reference. It should be a snippet though. I think this is because this is the one case where elixir sense offers a struct.

**Comment by scohen** [comment]
> There is a bug when you want to complete the child module

Yep, that's a bug. Can you verify that it only happens when completing the top level module with no dots? For example, `%Lexical|` shoes the bug, but `%Lexical.D|` works.

**Comment by scohen** [comment]
So right now, struct completions only offer suggestions for the immediate children of a module. Would it be better if they offered suggestions for all descendants (children, grandchildren, great grandchildren, etc)?

**Comment by scottming** [comment]
> if there's a leading %, no module completions are offered, this is intentional, since you're in a struct reference. It should be a snippet though. I think this is because this is the one case where elixir sense offers a struct.

not a snippet.

**Comment by scottming** [comment]
> Yep, that's a bug. Can you verify that it only happens when completing the top level module with no dots? For example, %Lexical| shoes the bug, but %Lexical.D| works.

*     `%Lexical.Document.Store.St|` -> `    %Lexical.Document.Store.Store{|}` works
* `%Lexical.Docu` -> `%Lexical.Document{}|` There are cursor and grandson module issues.

**Comment by scottming** [comment]
> So right now, struct completions only offer suggestions for the immediate children of a module. Would it be better if they offered suggestions for all descendants (children, grandchildren, great grandchildren, etc)?

right, And I think `%struc|` completions only offering struct is better, ignoring functions, fields, and so on.

**Comment by scottming** [comment]
Should we also keep the icon of the snippet consistent? Its normal icon looks like this:

image

**Comment by scohen** [comment]
The kind is based on the type of suggestion, not whether it's a snippet. All the items there are macros

**Comment by scohen** [comment]
@scottming, I couldn't tell from your response, but do you think it's better to return all descendant structs, or just the immediate children?

**Review by scottming — APPROVED** [review]

**Pull request merged**